### DISCUSSION TITLE
sqlalchemy + asyncpg still blocking in Fastapi Program

### DISCUSSION THREAD

**Discussion by mydu27** [sqlalchemy/sqlalchemy#9719] [Usage Questions]
### Describe the bug

I use sqlalchemy + asyncpg in my Fastapi progra, and I have a table with 1M data called userRecords. I noticed that If I send two request in the same time, the fastapi program will blocking until one request return. What wrong with it?

### SQLAlchemy Version in Use

1.4.41

### DBAPI (i.e. the database driver)

asyncpg

### Database Vendor and Major Version

postgresql 9.4

### Python Version

python 3.7

### Operating system

windows

### To Reproduce

```python
@router.get("/test-block")
async def test2(*, db: Session = Depends(get_db),
                request: Request) -> ApiResponse:
    """
    test block
    """
    data = await test(db=db, userId=5)
    return ApiResponse(msg='successful', data=data)

async def test(db: Session, *, userId: int) -> UserRecords:
    logger.info('start')
    stmt = select(UserRecords).where(UserRecords.userId == userId).order_by(UserRecords.createTime.desc())
    result = await db.execute(stmt)
    user_obj = result.scalars().first()
    logger.info('end')
    return user_obj
```

### Error

**db.execute() will spend at least five seconds. So when I send two request in the same time, my expect input in console is:** 
```
start
start
end
end
``` 
But it actually returned:
``` 
start
end
start
end
``` 

Later I realized my code make some mistakes. I should use limit(1) after order_by to improving the db.execute() speed. But I still want to know why this function is synchronously not async？

**Comment by CaselIT** [comment]
Hi,

I think the issue is that the application is busy loading 1m records. Try instead running `result = await db.execute(stmt.limit(1))`

The db io is not sync, but loading and processing all the records happens in python and there is no way of making it asynchronous

**Comment by mydu27** [comment]
> 你好，
> 
> 我认为问题在于应用程序正忙于加载 1m 记录。试着跑步`result = await db.execute(stmt.limit(1))`
> 
> db io 不同步，但加载和处理所有记录发生在 python 中，无法使其异步

Hi, Thanks. I already know `result = await db.execute(stmt.limit(1)) ` can solve this problem with a fast query.

But I noticed if `await db.execute` will cost several seconds, get_db function will blocking in the second request. I want to know why I can't get session in second request, until the `await db.execute()`  over in first request?

this is my get_db function:

```
Base = declarative_base()

engine = create_async_engine(
    settings.SQLALCHEMY_DATABASE_URI,
    future=True,
    echo=False,
)
SessionLocal = sessionmaker(expire_on_commit=False,
                            autoflush=False,
                            class_=AsyncSession)
SessionLocal.configure(bind=engine)

async def get_db() -> AsyncGenerator:
    async with SessionLocal() as db:
        db = db
    try:
        yield db
    except SQLAlchemyError as sql_ex:
        await db.rollback()
        raise sql_ex
    except HTTPException as http_ex:
        await db.rollback()
        raise http_ex
    finally:
        await db.close()

```
I think SessionLocal() is a non-blocking operation, where is wrong with my code?

**Comment by CaselIT** [comment]
As mentioned, if python is busy processing 1M rows that operation is sincronius. There is no IO going on there, so all other coroutines have to wait until that has ended.
